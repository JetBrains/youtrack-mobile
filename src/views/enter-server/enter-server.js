/* @flow */

import {
  ActivityIndicator,
  Image,
  Linking,
  ScrollView,
  Text,
  TextInput,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View
} from 'react-native';
import React, {Component} from 'react';
import {connect} from 'react-redux';
import {logo, IconBack} from '../../components/icon/icon';
import KeyboardSpacer from 'react-native-keyboard-spacer';
import usage from '../../components/usage/usage';
import {VERSION_DETECT_FALLBACK_URL} from '../../components/config/config';
import log from '../../components/log/log';
import clicksToShowCounter from '../../components/debug-view/clicks-to-show-counter';
import {resolveErrorMessage} from '../../components/error/error-resolver';
import type {AppConfigFilled} from '../../flow/AppConfig';
import {connectToNewYoutrack, openDebugView} from '../../actions/app-actions';
import throttle from 'lodash.throttle';
import {NETWORK_PROBLEM_TIPS} from '../../components/error-message/error-text-messages';

import ErrorMessageInline from '../../components/error-message/error-message-inline';
import {View as AnimatedView} from 'react-native-animatable';
import {UNIT} from '../../components/variables/variables';

import styles from './enter-server.styles';
import {formStyles} from '../../components/common-styles/form';

const CATEGORY_NAME = 'Choose server';
const protocolRegExp = /^https?:/i;
const CLOUD_DOMAIN = 'myjetbrains.com';

type Props = {
  serverUrl: string,
  connectToYoutrack: (newServerUrl: string) => Promise<AppConfigFilled>,
  onShowDebugView: Function,
  onCancel: () => any
};

type State = {
  serverUrl: string,
  connecting: boolean,
  error: ?string
};

const hitSlop = {top: UNIT, bottom: UNIT, left: UNIT, right: UNIT};

export class EnterServer extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      serverUrl: props.serverUrl,
      connecting: false,
      error: null
    };

    usage.trackScreenView(CATEGORY_NAME);
    log.info('Entering server URL view has been opened');
  }

  getPossibleUrls(enteredUrl: string) {
    if (protocolRegExp.test(enteredUrl)) {
      if (enteredUrl.indexOf('http:') === 0 && enteredUrl.indexOf(CLOUD_DOMAIN) !== -1) {
        enteredUrl = enteredUrl.replace('http:', 'https:');
        log.info('HTTP protocol was replaced for cloud instance', enteredUrl);
      }
      return [enteredUrl, `${enteredUrl}/youtrack`, `${enteredUrl}${VERSION_DETECT_FALLBACK_URL}`];
    }

    return [
      `https://${enteredUrl}`,
      `https://${enteredUrl}/youtrack`,
      `http://${enteredUrl}`,
      `http://${enteredUrl}/youtrack`,
      `http://${enteredUrl}${VERSION_DETECT_FALLBACK_URL}`
    ];
  }

  async onApplyServerUrlChange() {
    if (!this.isValidInput()) {
      return;
    }
    this.setState({connecting: true, error: null});
    const trimmedUrl = this.state.serverUrl.trim().replace(/\/$/i, '');

    const urlsToTry = this.getPossibleUrls(trimmedUrl);
    log.log(`Entered: "${this.state.serverUrl}", will try that urls: ${urlsToTry.join(', ')}`);

    let errorToShow = null;

    // eslint-disable-next-line no-unused-vars
    for (const url of urlsToTry) {
      log.log(`Trying: "${url}"`);
      try {
        await this.props.connectToYoutrack(url);
        log.log(`Successfully connected to ${url}`);
        return;
      } catch (error) {
        log.log(`Failed to connect to ${url}`, error);
        log.log(`Connection error for ${url}: ${error && error.toString()}`);
        if (error && error.isIncompatibleYouTrackError) {
          errorToShow = error;
          break;
        }
        errorToShow = errorToShow || error;
      }
    }

    const errorMessage = await resolveErrorMessage(errorToShow);
    this.setState({error: errorMessage, connecting: false});
  }

  isValidInput() {
    return throttle(() => {
      const url = (this.state.serverUrl || '').trim();
      return url.length > 0 && !url.match(/@/g);
    }, 500)();
  }

  render() {
    const {onShowDebugView, onCancel} = this.props;
    const isDisabled = this.state.connecting || !this.isValidInput();

    return (
      <ScrollView
        testID="enterServer"
        keyboardShouldPersistTaps="handled"
        keyboardDismissMode="on-drag"
        contentContainerStyle={styles.scrollContainer}
      >
        <View style={styles.container}>
          {onCancel && (
            <TouchableOpacity
              testID="enterServerBackButton"
              onPress={onCancel}
              style={styles.backIconButton}
            >
              <IconBack/>
            </TouchableOpacity>
          )}

          <View style={styles.formContent}>
            <TouchableWithoutFeedback
              testID="enterServerLogo"
              onPress={() => clicksToShowCounter(onShowDebugView)}
            >
              <Image style={styles.logoImage} source={logo}/>
            </TouchableWithoutFeedback>

            <View testID="enterServerHint">
              <Text style={styles.title}>Enter YouTrack URL</Text>
            </View>

            <TextInput
              testID="server-url"
              style={styles.input}
              autoCapitalize="none"
              autoFocus={true}
              selectTextOnFocus={true}
              autoCorrect={false}
              placeholder="youtrack-server.com:PORT"
              returnKeyType="done"
              keyboardType="url"
              underlineColorAndroid="transparent"
              onSubmitEditing={() => this.onApplyServerUrlChange()}
              value={this.state.serverUrl}
              onChangeText={(serverUrl) => this.setState({serverUrl})}/>

            {Boolean(this.state.error) && (
              <AnimatedView
                animation="slideInDown"
                duration={500}
                useNativeDriver
              >
                <ErrorMessageInline
                  testID="enterServerError"
                  error={this.state.error}
                  tips={NETWORK_PROBLEM_TIPS}
                />
              </AnimatedView>
            )}

            <TouchableOpacity
              style={[formStyles.button, isDisabled ? formStyles.buttonDisabled : null]}
              disabled={isDisabled}
              testID="next"
              onPress={() => this.onApplyServerUrlChange()}>
              <Text style={formStyles.buttonText}>Next</Text>
              {this.state.connecting && <ActivityIndicator style={styles.progressIndicator}/>}
            </TouchableOpacity>

          </View>

          <View
            testID="enterServerSupportLink"
            style={styles.supportLinkContent}
          >
            <TouchableOpacity
              hitSlop={hitSlop}
              onPress={() => Linking.openURL('https://youtrack-support.jetbrains.com/hc/en-us/requests/new')}
            >
              <Text style={formStyles.link}>Contact support</Text>
            </TouchableOpacity>
          </View>

          <KeyboardSpacer/>
        </View>
      </ScrollView>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  return {...ownProps};
};

const mapDispatchToProps = (dispatch) => {
  return {
    connectToYoutrack: newURL => dispatch(connectToNewYoutrack(newURL)),
    onShowDebugView: () => dispatch(openDebugView())
  };
};

// Needed to have a possibility to override callback by own props
const mergeProps = (stateProps, dispatchProps) => {
  return {
    ...dispatchProps,
    ...stateProps
  };
};

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(EnterServer);
