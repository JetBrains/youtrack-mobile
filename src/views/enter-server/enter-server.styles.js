import EStyleSheet from 'react-native-extended-stylesheet';
import {UNIT} from '../../components/variables/variables';
import {rowFormStyles} from '../../components/common-styles/form';
import {loginStylesForm} from '../../components/common-styles/login-form';
import {mainText} from '../../components/common-styles/typography';


export default EStyleSheet.create({
  ...loginStylesForm,

  container: {
    ...loginStylesForm.container,
    maxWidth: '100%',
    backgroundColor: '$background',
  },
  input: {
    ...rowFormStyles.input,
    marginTop: UNIT * 3,
    marginBottom: UNIT * 2,
    color: '$text',
    backgroundColor: '$boxBackground',
  },
  button: {
    ...rowFormStyles.button,
    maxWidth: 500,
  },
  title: {
    ...loginStylesForm.title,
    color: '$text',
  },
  errorContainer: {
    flexDirection: 'row',
    alignItems: 'flex-start',
    marginTop: UNIT * 2,
  },
  errorText: {
    flexGrow: 1,
  },
  infoIcon: {
    flexGrow: 0,
    marginTop: UNIT / 2,
    marginLeft: UNIT * 2,
  },
  text: {
    ...mainText,
    color: '$text',
  },
  placeholder: {
    color: '$icon',
  },
});
