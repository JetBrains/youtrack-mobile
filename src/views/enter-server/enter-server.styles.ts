import EStyleSheet from 'react-native-extended-stylesheet';
import {UNIT} from 'components/variables/variables';
import {rowFormStyles} from 'components/common-styles/form';
import {loginStylesForm} from 'components/common-styles/login-form';
import {mainText} from 'components/common-styles/typography';
const buttonMaxWidth = 500;
export default EStyleSheet.create({
  ...loginStylesForm,
  container: {...loginStylesForm.container},
  scrollContainer: {
    ...loginStylesForm.scrollContainer,
    alignItems: 'center',
    justifyContent: 'center',
  },
  input: {
    ...rowFormStyles.input,
    marginTop: UNIT * 3,
    marginBottom: UNIT * 2,
    color: '$text',
    backgroundColor: '$boxBackground',
  },
  button: {...rowFormStyles.button, maxWidth: buttonMaxWidth},
  title: {...loginStylesForm.title, color: '$text', textAlign: 'center'},
  errorContainer: {
    flexDirection: 'row',
    flex: 1,
    alignSelf: 'flex-start',
    maxWidth: buttonMaxWidth,
    marginTop: UNIT * 2,
    marginRight: UNIT * 2,
  },
  error: {
    maxWidth: '95%',
  },
  infoIcon: {
    marginTop: UNIT / 2,
    marginLeft: UNIT * 2,
  },
  text: {...mainText, color: '$text'},
  placeholder: {
    color: '$icon',
  },
});
