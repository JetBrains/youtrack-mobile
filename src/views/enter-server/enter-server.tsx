import {
  ActivityIndicator,
  Image,
  Linking,
  ScrollView,
  Text,
  TextInput,
  TouchableOpacity,
  TouchableWithoutFeedback,
  View,
} from 'react-native';
import React, {Component} from 'react';
import {connect} from 'react-redux';
import clicksToShowCounter from 'components/debug-view/clicks-to-show-counter';
import ErrorMessageInline from 'components/error-message/error-message-inline';
import IconMaterial from 'react-native-vector-icons/MaterialCommunityIcons';
import KeyboardSpacer from 'react-native-keyboard-spacer';
import log from 'components/log/log';
import Popup from 'components/popup/popup';
import usage from 'components/usage/usage';
import {ANALYTICS_ENTER_SERVER_PAGE} from 'components/analytics/analytics-ids.ts';
import {connectToNewYoutrack, openDebugView} from 'actions/app-actions';
import {formStyles} from 'components/common-styles/form';
import {HIT_SLOP} from 'components/common-styles';
import {i18n} from 'components/i18n/i18n';
import {IconBack, logo} from 'components/icon/icon';
import {NETWORK_PROBLEM_TIPS} from 'components/error-message/error-text-messages';
import {resolveErrorMessage} from 'components/error/error-resolver';
import {ThemeContext} from 'components/theme/theme-context';

import styles from './enter-server.styles';

import type {AppConfig} from 'types/AppConfig';
import type {AppState} from 'reducers';
import type {ReduxThunkDispatch} from 'types/Redux.ts';
import type {Theme, UIThemeColors} from 'types/Theme';

const protocolRegExp = /^http(s?):\/\//i;
const CLOUD_DOMAINS: string[] = ['myjetbrains.com', 'youtrack.cloud'];

interface Props {
  serverUrl: string;
  connectToYoutrack: (newServerUrl: string) => Promise<AppConfig>;
  onShowDebugView: (actionToPerform?: () => void, message?: string, numberOfTaps?: number) => void;
  onCancel: () => void;
  error?: string;
}

interface State {
  serverUrl: string;
  connecting: boolean;
  error: string | null;
  isErrorInfoModalVisible: boolean;
}

export class EnterServer extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      serverUrl: props.serverUrl,
      connecting: false,
      error: props.error || null,
      isErrorInfoModalVisible: false,
    };
    usage.trackScreenView(ANALYTICS_ENTER_SERVER_PAGE);
    log.info('Entering server URL view has been opened');
  }

  getPossibleUrls(enteredUrl: string): string[] {
    const targetURL: string = enteredUrl.toLowerCase();
    const isCloudURL: boolean = CLOUD_DOMAINS.some(
      (it: string) => targetURL.indexOf(`.${it}`) !== -1,
    );
    let urls: string[];

    if (isCloudURL) {
      const url: string = targetURL.replace(protocolRegExp, '');
      urls = [`https://${url}`, `https://${url}/youtrack`];
    } else {
      if (!protocolRegExp.test(targetURL)) {
        urls = [
          `https://${targetURL}`,
          `https://${targetURL}/youtrack`,
          `http://${targetURL}`,
          `http://${targetURL}/youtrack`,
        ];
      } else {
        urls = [targetURL, `${targetURL}/youtrack`];
      }
    }

    return urls;
  }

  async onApplyServerUrlChange(): Promise<string | null | undefined> {
    if (!this.isValidInput()) {
      return;
    }

    this.setState({
      connecting: true,
      error: null,
    });
    const trimmedUrl = this.state.serverUrl.trim().replace(/\/$/i, '');
    const urlsToTry = this.getPossibleUrls(trimmedUrl);
    log.log(
      `Entered: "${this.state.serverUrl}", will try that urls: ${urlsToTry.join(
        ', ',
      )}`,
    );
    let errorToShow = null;

    for (const url of urlsToTry) {
      log.log(`Trying: "${url}"`);

      try {
        await this.props.connectToYoutrack(url);
        log.log(`Successfully connected to ${url}`);
        return;
      } catch (error) {
        log.log(`Failed to connect to ${url}`, error);
        log.log(`Connection error for ${url}: ${error && error.toString()}`);

        if (error?.isIncompatibleYouTrackError) {
          errorToShow = error;
          break;
        }

        errorToShow = errorToShow || error;
      }
    }

    const errorMessage = await resolveErrorMessage(errorToShow);
    this.setState({
      error: errorMessage,
      connecting: false,
    });
    return errorMessage;
  }

  isValidInput(): any {
    const url = (this.state.serverUrl || '').trim();
    return url.length > 0 && !url.match(/@/g);
  }

  renderErrorInfoModalContent(): React.ReactNode {
    return (
      <React.Fragment>
        {NETWORK_PROBLEM_TIPS.map((tip: string, index: number) => {
          return (
            <Text
              key={`errorInfoTips-${index}`}
              style={styles.text}
            >{`${tip}\n`}</Text>
          );
        })}
      </React.Fragment>
    );
  }

  toggleErrorInfoModalVisibility: () => void = () => {
    const {isErrorInfoModalVisible} = this.state;
    this.setState({
      isErrorInfoModalVisible: !isErrorInfoModalVisible,
    });
  };

  render(): React.ReactNode {
    const {onShowDebugView, onCancel} = this.props;
    const {error, connecting, serverUrl, isErrorInfoModalVisible} = this.state;
    const isDisabled = connecting || !this.isValidInput();
    return (
      <ThemeContext.Consumer>
        {(theme: Theme) => {
          const uiThemeColors: UIThemeColors = theme.uiTheme.colors;
          return (
            <ScrollView
              testID="test:id/enterServer"
              keyboardShouldPersistTaps="handled"
              keyboardDismissMode="on-drag"
              style={styles.background}
              contentContainerStyle={styles.scrollContainer}
            >
              <View style={styles.container}>
                <View style={styles.backIconButtonContainer}>
                  {onCancel && (
                    <TouchableOpacity
                      testID="test:id/enterServerBackButton"
                      onPress={onCancel}
                      style={styles.backIconButton}
                    >
                      <IconBack />
                    </TouchableOpacity>
                  )}
                </View>

                <View style={styles.formContent}>
                  <TouchableWithoutFeedback
                    testID="enterServerLogo"
                    onPress={() => clicksToShowCounter(onShowDebugView)}
                  >
                    <Image style={styles.logoImage} source={logo} />
                  </TouchableWithoutFeedback>

                  <View testID="test:id/enterServerHint">
                    <Text style={styles.title}>
                      {i18n(
                        'Enter the web address for a YouTrack installation where you have a registered account',
                      )}
                    </Text>
                  </View>

                  <TextInput
                    testID="test:id/server-url"
                    accessibilityLabel="server-url"
                    accessible={true}
                    style={styles.input}
                    autoCapitalize="none"
                    autoFocus={true}
                    selectTextOnFocus={true}
                    autoCorrect={false}
                    placeholder="my-youtrack-server.com"
                    placeholderTextColor={styles.placeholder.color}
                    returnKeyType="done"
                    keyboardType="url"
                    underlineColorAndroid="transparent"
                    onSubmitEditing={() => this.onApplyServerUrlChange()}
                    value={serverUrl}
                    onChangeText={serverUrl =>
                      this.setState({
                        serverUrl,
                        error: null,
                      })
                    }
                  />

                  <TouchableOpacity
                    style={[
                      formStyles.button,
                      isDisabled ? formStyles.buttonDisabled : null,
                    ]}
                    disabled={isDisabled}
                    testID="test:id/next"
                    accessibilityLabel="next"
                    accessible={true}
                    onPress={() => this.onApplyServerUrlChange()}
                  >
                    <Text
                      style={[
                        formStyles.buttonText,
                        isDisabled && formStyles.buttonTextDisabled,
                      ]}
                    >
                      {i18n('Next')}
                    </Text>
                    {connecting && (
                      <ActivityIndicator style={styles.progressIndicator} />
                    )}
                  </TouchableOpacity>

                  {Boolean(error) && (
                    <View style={styles.errorContainer}>
                      <ErrorMessageInline style={styles.error} error={error} />

                      <TouchableOpacity
                        style={styles.infoIcon}
                        hitSlop={HIT_SLOP}
                        onPress={this.toggleErrorInfoModalVisibility}
                      >
                        <IconMaterial
                          name="information"
                          size={24}
                          color={uiThemeColors.$iconAccent}
                        />
                      </TouchableOpacity>
                    </View>
                  )}
                </View>

                <View
                  testID="test:id/enterServerHelpLink"
                  style={styles.supportLinkContent}
                >
                  <TouchableOpacity
                    hitSlop={HIT_SLOP}
                    onPress={() =>
                      Linking.openURL(
                        'https://www.jetbrains.com/help/youtrack/incloud/youtrack-mobile.html#start-using-youtrack-mobile',
                      )
                    }
                  >
                    <Text style={formStyles.link}>{i18n('Get help')}</Text>
                  </TouchableOpacity>
                </View>

                <View
                  testID="enterServerSupportLink"
                  style={styles.supportLinkContent}
                >
                  <TouchableOpacity
                    hitSlop={HIT_SLOP}
                    onPress={() =>
                      Linking.openURL(
                        'https://youtrack-support.jetbrains.com/hc/en-us/requests/new',
                      )
                    }
                  >
                    <Text style={formStyles.link}>
                      {i18n('Contact support')}
                    </Text>
                  </TouchableOpacity>
                </View>

                {isErrorInfoModalVisible && (
                  <Popup
                    childrenRenderer={this.renderErrorInfoModalContent}
                    onHide={this.toggleErrorInfoModalVisibility}
                  />
                )}

                <KeyboardSpacer />
              </View>
            </ScrollView>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}

const mapStateToProps = (state: AppState, ownProps: Props) => {
  return {...ownProps};
};

type MappedActions = {
  onShowDebugView: () => void;
  connectToYoutrack: (url: string) => void;
};

const mapDispatchToProps = (dispatch: ReduxThunkDispatch): MappedActions => {
  return {
    connectToYoutrack: (url: string) => dispatch(connectToNewYoutrack(url)),
    onShowDebugView: () => dispatch(openDebugView()),
  };
};

// Needed to have a possibility to override callback by own props
const mergeProps = (stateProps: Props, dispatchProps: MappedActions): Props & MappedActions => {
  return {...dispatchProps, ...stateProps};
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
  mergeProps,
)(EnterServer);
