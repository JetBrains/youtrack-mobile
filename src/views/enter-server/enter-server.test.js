import React from 'react';

import {shallow} from 'enzyme';
import sinon from 'sinon';
import toJson from 'enzyme-to-json';

import {EnterServer} from './enter-server';

xdescribe('EnterServer', () => {
  const serverUrl = 'http://example.com';
  let connectToYouTrack;
  let onCancel;
  let wrapper;
  let waitForNextTick;
  let connectPromise;

  function renderComponent(url = serverUrl) {
    wrapper = shallow(<EnterServer serverUrl={url} connectToYoutrack={connectToYouTrack} onCancel={onCancel}/>);
  }

  beforeEach(() => {
    connectPromise = Promise.resolve({foo: 'bar'});
    waitForNextTick = () => new Promise(resolve => setTimeout(resolve));

    connectToYouTrack = sinon.spy(() => connectPromise);
    onCancel = sinon.spy();

    renderComponent();
  });

  describe('Render', () => {

    it('should render', () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render element', () => {
      expect(findByTestId('enterServer')).toHaveLength(1);
      expect(findByTestId('enterServerBackButton')).toHaveLength(1);
      expect(findByTestId('server-url')).toHaveLength(1);
      expect(findByTestId('enterServerLogo')).toHaveLength(1);
      expect(findByTestId('next')).toHaveLength(1);
      expect(findByTestId('enterServerHint')).toHaveLength(1);
      expect(findByTestId('enterServerSupportLink')).toHaveLength(1);
    });
  });

  it('should connect to server', async () => {
    const connectButton = wrapper.find({testID: 'next'});
    connectButton.simulate('press');
    await waitForNextTick();

    connectToYouTrack.should.have.been.calledWith(serverUrl);
  });

  it('should add protocol for entered URL', async () => {
    renderComponent('foo.bar');
    wrapper.find({testID: 'next'}).simulate('press');
    await waitForNextTick();

    connectToYouTrack.should.have.been.calledWith('https://foo.bar');
  });

  it('should replace HTTP with HTTPS for a cloud instance', async () => {
    renderComponent('http://foo.myjetbrains.com');
    wrapper.find({testID: 'next'}).simulate('press');
    await waitForNextTick();

    connectToYouTrack.should.have.been.calledWith('https://foo.myjetbrains.com');
  });

  it('should strip wrapping spaces', async () => {
    renderComponent('   foo.bar ');
    wrapper.find({testID: 'next'}).simulate('press');
    await waitForNextTick();

    connectToYouTrack.should.have.been.calledWith('https://foo.bar');
  });

  it('should strip tailing slash', async () => {
    renderComponent('http://foo.bar/');
    wrapper.find({testID: 'next'}).simulate('press');
    await waitForNextTick();

    connectToYouTrack.should.have.been.calledWith('http://foo.bar');
  });

  it('should try next URL on failure if protocol is entered', async () => {
    connectPromise = Promise.reject({message: 'test reject'});
    const connectButton = wrapper.find({testID: 'next'});

    connectButton.simulate('press');
    await waitForNextTick();
    connectButton.simulate('press');
    await waitForNextTick();

    connectToYouTrack.should.have.been.calledWith(serverUrl);
    connectToYouTrack.should.have.been.calledWith(`${serverUrl}/youtrack`);
    connectToYouTrack.should.have.been.calledWith(`${serverUrl}/rest/workflow/version`);
  });


  it('should try next URL on failure if no protocol entered', async () => {
    connectPromise = Promise.reject({message: 'test reject'});
    renderComponent('foo.bar');
    const connectButton = wrapper.find({testID: 'next'});

    connectButton.simulate('press');
    await waitForNextTick();
    connectButton.simulate('press');
    await waitForNextTick();
    connectButton.simulate('press');
    await waitForNextTick();
    connectButton.simulate('press');
    await waitForNextTick();

    connectToYouTrack.should.have.been.calledWith('https://foo.bar');
    connectToYouTrack.should.have.been.calledWith('https://foo.bar/youtrack');
    connectToYouTrack.should.have.been.calledWith('http://foo.bar');
    connectToYouTrack.should.have.been.calledWith('http://foo.bar/youtrack');
    connectToYouTrack.should.have.been.calledWith('http://foo.bar/rest/workflow/version');
  });

  it('should stop and display error if `IncompatibleYouTrackError` is thrown', async () => {
    const incompatibleYoutrackMsg = 'Incompatible youtrack';
    connectPromise = Promise.reject({
      isIncompatibleYouTrackError: true,
      message: incompatibleYoutrackMsg
    });

    wrapper.find({testID: 'next'}).simulate('press');
    await waitForNextTick();

    wrapper.state('error').should.equal(incompatibleYoutrackMsg);
  });

  it('should not allow empty input', () => {
    const instance = shallow(
      <EnterServer serverUrl={''} connectToYoutrack={connectToYouTrack} onCancel={onCancel}/>
    ).instance();
    instance.isValidInput().should.be.false;
  });

  it('should not allow `@` character in server input (to not confuse users with email)', () => {
    const instance = shallow(
      <EnterServer serverUrl={'foo@bar.com'} connectToYoutrack={connectToYouTrack} onCancel={onCancel}/>
    ).instance();

    instance.isValidInput().should.be.false;
  });

  it('should allow not empty input', () => {
    const instance = shallow(
      <EnterServer serverUrl={'someserver'} connectToYoutrack={connectToYouTrack} onCancel={onCancel}/>
    ).instance();

    instance.isValidInput().should.be.true;
  });


  function findByTestId(testId) {
    return wrapper && wrapper.find({testID: testId});
  }
});
