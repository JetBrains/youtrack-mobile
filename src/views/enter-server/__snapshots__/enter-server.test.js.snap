// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnterServer Render should render 1`] = `
<ScrollViewMock
  contentContainerStyle={
    Object {
      "flexGrow": 1,
    }
  }
  keyboardDismissMode="on-drag"
  keyboardShouldPersistTaps="handled"
  testID="enterServer"
>
  <View
    style={
      Object {
        "backgroundColor": "#FFF",
        "flexDirection": "column",
        "flexGrow": 1,
        "justifyContent": "space-between",
        "paddingLeft": 32,
        "paddingRight": 32,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
      style={
        Object {
          "flexGrow": 0,
          "marginLeft": -32,
          "padding": 16,
          "width": 72,
        }
      }
      testID="enterServerBackButton"
    >
      <IconBack />
    </TouchableOpacity>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginTop": 16,
        }
      }
    >
      <TouchableWithoutFeedback
        onPress={[Function]}
        testID="enterServerLogo"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/components/icon/youtrack-logo-512.png",
            }
          }
          style={
            Object {
              "height": 80,
              "resizeMode": "contain",
            }
          }
        />
      </TouchableWithoutFeedback>
      <View
        testID="enterServerHint"
      >
        <Text
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "500",
              "marginTop": 16,
              "textAlign": "center",
            }
          }
        >
          Enter YouTrack URL
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        autoFocus={true}
        keyboardType="url"
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="youtrack-server.com:PORT"
        rejectResponderTermination={true}
        returnKeyType="done"
        selectTextOnFocus={true}
        style={
          Object {
            "backgroundColor": "#f6f6f6",
            "borderRadius": 8,
            "color": "#000",
            "fontSize": 16,
            "marginBottom": 16,
            "marginTop": 24,
            "padding": 12,
            "paddingLeft": 8,
            "paddingRight": 8,
            "width": "100%",
          }
        }
        testID="server-url"
        underlineColorAndroid="transparent"
        value="http://example.com"
      />
      <TouchableOpacity
        activeOpacity={0.2}
        disabled={false}
        onPress={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#FE0082",
              "borderRadius": 8,
              "padding": 12,
              "width": "100%",
            },
            null,
          ]
        }
        testID="next"
      >
        <Text
          style={
            Object {
              "alignSelf": "center",
              "color": "#FFF",
              "fontSize": 16,
              "textAlign": "center",
            }
          }
        >
          Next
        </Text>
      </TouchableOpacity>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "marginBottom": 16,
          "marginTop": 16,
        }
      }
      testID="enterServerSupportLink"
    >
      <TouchableOpacity
        activeOpacity={0.2}
        hitSlop={
          Object {
            "bottom": 8,
            "left": 8,
            "right": 8,
            "top": 8,
          }
        }
        onPress={[Function]}
      >
        <Text
          style={
            Object {
              "color": "#FE0082",
            }
          }
        >
          Contact support
        </Text>
      </TouchableOpacity>
    </View>
    <KeyboardSpacer
      onToggle={[Function]}
      topSpacing={0}
    />
  </View>
</ScrollViewMock>
`;
